### PR TITLE
implement garbage collection reconcilation

### PR DESCRIPTION
**Pull request by elenz97**
- add logic for garbage collection job creation & updates
- add a mock garbage collection to the testing pkg's `CreateInstance` function
- misc documentation changes

### CODE REVIEW

File: examples/README.md [hunk 1]
```diff
@@ -57,6 +61,11 @@ spec:
 Note: Specyfing an empty string for the `harborAdminPassword`-key in `spec.helmChart.valuesYaml` will trigger password generation by the harbor instance itself.
 The admin password will then be saved under the key `HARBOR_ADMIN_PASSWORD` in a secret named `HELM_RELEASE_NAME`-`harbor-core`.
 
+[Harbor Garbage Collection](https://goharbor.io/docs/1.10/administration/garbage-collection/) can be configured via `spec.garbageCollection`.
+Valid values for `.spec.garbageCollection.schedule.type` are `Hourly`, `Daily`, `Weekly`, `Custom`, `Manual`, and `None` (each starting with a capital letter).
```
Comment (martin-helmich) [inline examples/README.md:65]: So... What is `garbageCollection.schedule.cron` even good for? Should we even expose it at all?

File: pkg/apis/registries/v1alpha1/instance_types.go [hunk 2]
```diff
@@ -56,6 +57,23 @@ type InstanceSpec struct {
...
+	Schedule *h.ScheduleParam `json:"schedule"`
+
+	// +optional
+	Name string `json:"name,omitempty"`
+	// +optional
+	Status string `json:"status,omitempty"`
+	// +optional
+	ID int64 `json:"id,omitempty"`
+	// +optional
+	Parameters map[string]byte `json:"parameters,omitempty"`
```
Comment (martin-helmich) [inline pkg/apis/registries/v1alpha1/instance_types.go:76]: A map of _single bytes_, really? Sure this wasn't supposed to be `map[string][]byte` or `map[string]string`?

File: pkg/controller/instance/garbage_collection.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		return err
+	}
+
+	// Assume that the garbage collection's Schedule is nil, if it has not been created yet
+	if gc.AdminJobSchedule == nil {
+		err = harborClient.System().CreateSystemGarbageCollectionSchedule(newGc)
+		if err != nil {
+			return err
+		}
+		return nil
```
Comment (martin-helmich) [inline pkg/controller/instance/garbage_collection.go:41]:
```suggestion
		return harborClient.System().CreateSystemGarbageCollectionSchedule(newGc)
```

File: pkg/controller/instance/garbage_collection.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+			return err
+		}
+	}
+
+	return nil
+}
+
+// BuildGarbageCollectionScheduleFromSpec constructs and returns a Harbor AdminJobReq from the spec
+func BuildGarbageCollectionScheduleFromSpec(harbor *registriesv1alpha1.Instance) (h.AdminJobReq, error) {
+	if harbor.Spec.GarbageCollection.Schedule == nil {
```
Comment (martin-helmich) [inline pkg/controller/instance/garbage_collection.go:57]: Maybe also test for `harbor.Spec.GarbageCollection == nil`? Or is that already done someplace else?

File: pkg/controller/instance/garbage_collection.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		return nil
+	case h.ScheduleTypeDaily:
+		return nil
+	case h.ScheduleTypeWeekly:
+		return nil
+	case h.ScheduleTypeManual:
+		return nil
+	case h.ScheduleTypeNone:
+		return nil
+	}
```
Comment (martin-helmich) [inline pkg/controller/instance/garbage_collection.go:103]:
```suggestion
	switch param.Type {
	case h.ScheduleTypeCustom, h.ScheduleTypeHourly, h.ScheduleTypeDaily, h.ScheduleTypeWeekly, h.ScheduleTypeManual, h.ScheduleTypeNone:
		return nil
	default:
		return fmt.Errorf("the provided garbage collection schedule type is invalid %s", param.Type)
	}
```

File: pkg/controller/instance/garbage_collection.go (hunk 4 above)
Comment (martin-helmich) [reply pkg/controller/instance/garbage_collection.go:57]: OK, I noticed that that condition is already checked in `ReconcileInstance.Reconcile()`; but this method is public (does it need to be?), so there may be other code paths leading into this function (now or in the future).

File: pkg/controller/instance/instance_controller.go [hunk 6]
```diff
@@ -144,6 +144,13 @@ func (r *ReconcileInstance) Reconcile(request reconcile.Request) (reconcile.Resu
 			return r.patchInstance(ctx, originalInstance, harbor)
 		}
 
+		if harbor.Spec.GarbageCollection != nil {
+			err := r.reconcileGarbageCollection(ctx, harbor)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
```
Comment (martin-helmich) [inline pkg/controller/instance/instance_controller.go:151]:
```suggestion
			if err := r.reconcileGarbageCollection(ctx, harbor); err != nil {
				return reconcile.Result{}, err
			}
```

File: examples/README.md (hunk 1 above)
Comment (elenz97) [reply examples/README.md:65]: This **has** to be set to a cron expression. The Harbor API will return an error when leaving this out, sadly. Im going to add comments on this, here.

File: pkg/apis/registries/v1alpha1/instance_types.go (hunk 2 above)
Comment (elenz97) [reply pkg/apis/registries/v1alpha1/instance_types.go:76]: Should've been `map[string]string` :+1:

**Pull request merged**